file/reed_solomon_file.go

```diff
@@ -0,0 +1,466 @@
+
+// TODO: 12/11 possibly create a pool DP if not possible to have a session..
+// But use api.core().getSession(ctx)?
+
+const (
+	ReedSolomonDagOff   = 0
+	ReedSolomonDagOpen  = 1
+	ReedSolomonDagNext  = 2
+	ReedSolomonDagClose = 3
+)
+
+type ReedSolomonDag struct {
+	state     int
+	buff      *bytes.Buffer
+	offset    uint64
+	curFile   *uio.FileNode
+	lock      sync.RWMutex
+	cidString string
+}
+
+type rsDirectory struct {
+	ctx       context.Context
+	dserv     ipld.DAGService
+	dir       uio.ReedSolomonDirectory
+	size      int64
+	cidString string
+}
+
+func (d *rsDirectory) Close() error {
+	return nil
+}
+
+func (d *rsDirectory) Entries() files.DirIterator {
+
+	fileCh := make(chan interface{}, prefetchFiles)
+	errCh := make(chan error, 1)
+	// Invoke goroutine to provide links of the current receiver `d`
+	// via `fileCh`.
+	go func() {
+		errCh <- d.dir.ForEachLink(d.ctx, func(link interface{}) error {
+			if d.ctx.Err() != nil {
+				return d.ctx.Err()
+			}
+			select {
+			case fileCh <- link:
+			case <-d.ctx.Done():
+				return d.ctx.Err()
+			}
+			return nil
+		})
+
+		// close channels indicating sender side is done.
+		close(errCh)
+		close(fileCh)
+	}()
+
+	return &rsIterator{
+		ctx:       d.ctx,
+		cidString: d.cidString,
+		files:     fileCh,
+		rsDir:     d,
+		errCh:     errCh,
+		dserv:     d.dserv,
+	}
+}
+
+func (d *rsDirectory) Size() (int64, error) {
+	return d.size, nil
+}
+
+type rsIterator struct {
+	state     int
+	ctx       context.Context
+	cidString string
+	files     chan interface{}
+	dserv     ipld.DAGService
+	rsDir     *rsDirectory
+
+	curName string
+	curFile files.Node
+
+	err   error
+	errCh chan error
+}
+
+func (it *rsIterator) Name() string {
+	return it.curName
+}
+
+func (it *rsIterator) Node() files.Node {
+	return it.curFile
+}
+
+func (it *rsIterator) Next() bool {
+	if it.err != nil {
+		return false
+	}
+
+	var l interface{}
+	var ok bool
+	// Loop until getting `l` without an error.
+	for !ok { // while ok == false
+		if it.files == nil && it.errCh == nil {
+			return false
+		}
+		select {
+		case l, ok = <-it.files:
+			if !ok {
+				it.files = nil
+			}
+		case err := <-it.errCh:
+			it.errCh = nil
+			it.err = err
+
+			if err != nil {
+				return false
+			}
+		}
+	}
+
+	it.curFile = nil
+
+	node, err := GetRsNode(l)
+	if err != nil {
+		it.err = err
+		return false
+	}
+
+	it.curName = node.Name()
+	if it.curName == uio.SmallestString {
+		return it.err == nil
+	}
+
+	switch nd := node.(type) {
+	case *uio.DirNode:
+		it.curFile, it.err = NewReedSolomonSubDirectory(it.ctx, it.dserv, it.cidString, nd)
+	case *uio.FileNode:
+		it.curFile, it.err = NewReedSolomonFileUnderDirectory(it.ctx, it.dserv, it.cidString, nd)
+	case *uio.SymlinkNode:
+		it.curFile, it.err = files.NewLinkFile(nd.Data, nil), nil
+	default:
+		it.err = errors.New("unexpected Node at Next(), possibly program error")
+		return false
+	}
+
+	return it.err == nil
+}
+
+func GetRsNode(l interface{}) (uio.Node, error) {
```

Comment:
is there no other generic type to use here? Looks like you can just do some type of `json.Unmarshal` for a `BaseNode` type?

Manually unraveling map[string]interface{} is a lot more work than defining a struct and unmarshal in one go